File: HRMS/HRMS.Web/Modules/ServerTypes/Operations/LeaveBalanceForm.ts
import { DecimalEditor, EnumEditor, initFormType, IntegerEditor, LookupEditor, PrefixedContext } from "@serenity-is/corelib";
import { LeaveType } from "./LeaveType";

export interface LeaveBalanceForm {
    EmployeeId: LookupEditor;
    LeaveType: EnumEditor;
    Year: IntegerEditor;
    Allocated: DecimalEditor;
    Used: DecimalEditor;
}

export class LeaveBalanceForm extends PrefixedContext {
    static readonly formKey = 'Operations.LeaveBalance';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!LeaveBalanceForm.init) {
            LeaveBalanceForm.init = true;

            var w0 = LookupEditor;
            var w1 = EnumEditor;
            var w2 = IntegerEditor;
            var w3 = DecimalEditor;

            initFormType(LeaveBalanceForm, [
                'EmployeeId', w0,
                'LeaveType', w1,
                'Year', w2,
                'Allocated', w3,
                'Used', w3
            ]);
        }
    }
}

[LeaveType]; // referenced types

File: HRMS/HRMS.Web/Modules/Operations/LeaveBalance/LeaveBalanceForm.cs
using Serenity.ComponentModel;
using System.ComponentModel;

namespace HRMS.Operations.Forms;

[FormScript("Operations.LeaveBalance")]
[BasedOnRow(typeof(LeaveBalanceRow), CheckNames = true)]
public class LeaveBalanceForm
{
    public int EmployeeId { get; set; }
    public LeaveType LeaveType { get; set; }
    public int Year { get; set; }
    public decimal Allocated { get; set; }
    public decimal Used { get; set; }
}
